### PR TITLE
feat: rewrite the scanner in C

### PR DESCRIPTION
**Pull request by amaanq**

### CODE REVIEW

**Comment by 6cdh** [comment]
Thank you for submitting this PR.

However, I'm afraid I can't approve it.

Is there a good reason to rewrite with these cost:

* 2x more code than the old C++ implementation
* many awkward C macros that could be replaced by function
* a hand-written macro string implementation

**Comment by amaanq** [comment]
Well, one reason is that I make use of the serialize/deserialize functions to load data that might change from editing/incremental changes (this could also be done in the C++ code too, so "more code" isn't a valid reason imo)

The main reason is portability and ease of compilation, so that projects using a pure C toolchain can use it easily.

**Comment by 6cdh** [comment]
> Well, one reason is that I make use of the serialize/deserialize functions to load data that might change from editing/incremental changes (this could also be done in the C++ code too, so "more code" isn't a valid reason imo)

The `scan` function only reads a token once.
The old C++ implementation doesn't need `create`/`serialize`/`deserialize`/`destroy` functions as it doesn't maintain states between read tokens.

While the new C implementation needs it because it uses `create` and `destroy` to allocate and free memory.
Am I right?

I think it's unnecessary complexity especially since these new functions are complex and hard to read and maintain.
All allocation and free can happen in the `scan` function and keep serialize functions clean.

> The main reason is portability and ease of compilation, so that projects using a pure C toolchain can use it easily.

I only saw these problems in neovim-treesitter issues. Is it a neovim-treesitter specific fix? (just curious)

It's a valid reason to improve this parser anyway. I will review code later.

File: src/scanner.c [hunk 1]
```diff
@@ -0,0 +1,200 @@
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+#include <tree_sitter/parser.h>
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
```
Comment (6cdh) [inline src/scanner.c:6]: Could you rewrite all these macros to functions?
like this:
```c
int max(int a, int b) {
    return a > b ? a : b;
}
```

The reason is macro don't have types annotations on parameters.
At the first glance, we don't know the type of parameters and the return value.
Another reason is the preprocessor simply text level expand code, it's very easy to make mistake.
I don't know the advantage of macros here. The code complexity and cognitive overhead of macro function is not necessary.

File: src/scanner.c [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+// Suppose terminator is `T`, newline (\n) is `$`,
+// It should accept "#<<T$T" or "#<<T$...$T", where `...`
+// is the string content.
+bool scan(Scanner *scanner, TSLexer *lexer, const bool *valid_symbols) {
+    if (!valid_symbols[HERE_STRING_BODY]) {
+        return false;
+    }
+
+    STRING_CLEAR(scanner->terminator_str);
```
Comment (6cdh) [inline src/scanner.c]: Consider allocate and free `scanner` struct here, and keep `tree_sitter_racket_external_scanner_*` functions empty.

Comment (6cdh) [reply src/scanner.c]: Also consider rename `scanner` to `here_string_states` or something like that.

**Comment by amaanq** [comment]
> The `scan` function only reads a token once. The old C++ implementation doesn't need `create`/`serialize`/`deserialize`/`destroy` functions as it doesn't maintain states between read tokens.
> 
> While the new C implementation needs it because it uses `create` and `destroy` to allocate and free memory. Am I right?

No, you too would've benefitted from using the serialize/deserialize functions. They are used during error recovery and incremental parsing (code editing) to pass instantaneous changes faster (iirc). The scanner would work just fine without it, but it's better to have it if you do store stateful data (like the heredoc delimiter).

It would benefit neovim and any other parser using a pure C toolchain - I've been rewriting core repo scanners for neovim to use and thought I might as well rewrite them all to keep it simpler and more portable.

File: src/scanner.c (hunk 1 above)
Comment (amaanq) [reply src/scanner.c:6]: That's fine, I can do that

File: src/scanner.c (hunk 2 above)
Comment (amaanq) [reply src/scanner.c]: I disagree because serialize/deserialize are called multiple times in between scans.

Scanner is also the common naming convention in any ts repo

**Comment by 6cdh** [comment]
> No, you too would've benefitted from using the serialize/deserialize functions. They are used during error recovery and incremental parsing (code editing) to pass instantaneous changes faster (iirc). The scanner would work just fine without it, but it's better to have it if you do store stateful data (like the heredoc delimiter).

I don't know the design of tree-sitter.
From the [tree-sitter#external-scanners docs](https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners)

> The function (serialize) is called every time the external scanner successfully recognizes a token.

When the `scan` successfully read a token, all its states (here string terminator or current line) are no longer useful.
Thus serialize function never really write useful data and also don't restore useful data.
That's why I say it doesn't have states.

File: src/scanner.c (hunk 2 above)
Comment (6cdh) [reply src/scanner.c]: What the `scanner` means here? We know the whole file is for scanner. The name doesn't reveal useful information.

The struct

```
typedef struct {
    bool valid;
    String terminator_str;
    String current_line;
} Scanner;
```

stores states that relates to here string parse not general scanner states.
I think `here_string_states` is a more precise name.

**Comment by amaanq** [comment]
Hmm that's fair, it's more useful for indents or items spanning multiple lines/calls to scan..however error recovery backtracks when a token *wasn't* recognized in any context - so in this case it *could* be useful. I'm not too familiar with the internals of that

File: src/scanner.c (hunk 1 above)
Comment (6cdh) [reply src/scanner.c:6]: could you also rewrite `MAX`?

File: src/scanner.c (hunk 2 above)
Comment (amaanq) [reply src/scanner.c]: > What the scanner means here? We know the whole file is for scanner. The name doesn't reveal useful information.

The memory allocated by the external scanner (scanner.c) for all the data/information it contains, that I (and every other ts-repo) find fitting. I think the field names give you the information needed for what's going on..but it is your repo and I can change it if you'd like

**Comment by 6cdh** [comment]
From this discussion: [tree-sitter repo issue 224 Error-recovery strategy](https://github.com/tree-sitter/tree-sitter/issues/224)

The error recovery is only work on token level, not sub token level.

For example once `scan` fails, we can informally say it reads the first half token successfully and the second half token fail.
But tree-sitter thinks it read 0 token.
If a user change the second half token to make correct code, then the first half information of `scan` will be useful.
However, the tree-sitter error recovery not work in this way.

So I still support keep (de)serialize functions empty in tree-sitter-racket.

File: src/scanner.c [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+#include <string.h>
+#include <tree_sitter/parser.h>
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+enum TokenType {
+    HERE_STRING_BODY,
+};
+
+typedef struct {
```
Comment (6cdh) [inline src/scanner.c:12]: consider write comment that points out the invariant of `String`: `cap == strlen(data) + 1`.
consider change all `uint32_t` to `size_t` as it can't work well on 64-bit machine.

File: src/scanner.c [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    uint32_t len;
+    uint32_t cap;
+    char *data;
+} String;
+
+static String string_new() {
+    return (String){.cap = 16, .len = 0, .data = calloc(1, sizeof(char) * 17)};
```
Comment (6cdh) [inline src/scanner.c]: consider write a comment that points out why we have 17 not 16.
Please process the case that `calloc` fails and return `NULL`. I think we can just `exit(EXIT_FAILURE);` in this case.

File: src/scanner.c [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+static void string_resize(String *vec, uint32_t cap) {
+    void *block = realloc(vec->data, (cap + 1) * sizeof(vec->data[0]));
+    assert(block != NULL);
```
Comment (6cdh) [inline src/scanner.c]: `assert` is a debug macro, can be removed by compiler in optimized release code,
I think explicitly `exit(EXIT_FAILURE);` is enough.

File: src/scanner.c [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    if (vec->cap < cap) {
+        string_resize(vec, cap);
+    }
+}
+
+static void string_push(String *vec, int32_t elem) {
+    if (vec->cap == vec->len) {
+        string_resize(vec, MAX(16, vec->len * 2));
+    }
+    vec->data[vec->len++] = (char)elem;
```
Comment (6cdh) [inline src/scanner.c]: there is a bug that `int32_t` character is converted to `char`.
`int32_t` can hold unicode character, while `char` only hold ASCII character.
I suggest push 4 times for each char in `int32_t`.

File: src/scanner.c [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+static void string_free(String *vec) {
+    if (vec->data != NULL) {
+        free(vec->data);
```
Comment (6cdh) [inline src/scanner.c:56]: consider set `vec->data` to `NULL` after free.

File: src/scanner.c [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    }
+}
+
+static void string_clear(String *vec) {
+    vec->len = 0;
+    memset(vec->data, 0, vec->cap * sizeof(char));
```
Comment (6cdh) [inline src/scanner.c]: There is a possible performance penalty, consider using `memset(vec->data, 0, vec->len)`.

**Review by 6cdh — CHANGES_REQUESTED** [review]

File: src/scanner.c (hunk 8 above)
Comment (amaanq) [reply src/scanner.c]: needs to be len+1 but sure

File: src/scanner.c (hunk 3 above)
Comment (amaanq) [reply src/scanner.c:12]: I don't understand the point of that

Comment (6cdh) [reply src/scanner.c:12]: My wrong. I misplaced this comment. See below.

File: src/scanner.c [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+enum TokenType {
+    HERE_STRING_BODY,
+};
+
+typedef struct {
+    uint32_t len;
+    uint32_t cap;
+    char *data;
+} String;
```
Comment (6cdh) [inline src/scanner.c:16]: The comment above should be placed here.

> consider write comment that points out the invariant of String: cap == strlen(data) + 1.

This is unnecessary as there are comment in `string_new` that points out there are a null character `\0`.

> change all uint32_t to size_t

`uint32_t` is always 32 bits, `size_t` is 32 bits on 32-bit platform, and 64 bits on 64-bit platform.
(It's not precise, see [cppreference - size_t](https://en.cppreference.com/w/c/types/size_t) and [cppreference - arithmetic types](https://en.cppreference.com/w/c/language/arithmetic_types))

`uint32_t` could overflow for very large source code (4GB+).

It may be crazy that someone has such a large source code file.
But it's possible.
A more important reason is that parser, is the basic component of a general programming language tool. We should not suppose what people use this language for and what this language is designed for. All we can do is let parser stable, correct, and reliable.

You need to change all variable that indicate the size of a sequence of bytes to `size_t`.

File: src/scanner.c [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+static void string_grow(String *vec, uint32_t cap) {
+    if (vec->cap < cap) {
+        string_resize(vec, cap);
+    }
+}
+
+static void string_push(String *vec, int32_t elem) {
+    if (vec->cap == vec->len) {
+        string_resize(vec, MAX(16, vec->len * 2));
+    }
```
Comment (6cdh) [inline src/scanner.c:49]: use

```
if (vec->len + sizeof(elem) >= vec->cap) {
    // ...
}
```

I think this is the last change.

File: src/scanner.c (hunk 9 above)
Comment (amaanq) [reply src/scanner.c:16]: with that rationale - wouldn't uint64_t be better

Comment (6cdh) [reply src/scanner.c:16]: See [Stackoverflow - What is size_t in C?](https://stackoverflow.com/questions/2550774/what-is-size-t-in-c)

TLDR. `size_t` is a type guaranteed to hold any array index. It scales very well.
On 32-bit platform, `uint64_t` is unnecessary large, and even the pointer size (or the address space size) is 32-bit..

Comment (6cdh) [reply src/scanner.c:16]: BTW, all these subtle details is why I didn't use C at the first time.

**Comment by 6cdh** [comment]
Looks good. I will change the code a little bit and update docs.

File: src/scanner.c (hunk 9 above)
Comment (amaanq) [reply src/scanner.c:16]: I don't particularly think such a check is really important...is it really rational to account for handling heredoc strings more than 4gb :shrug:

**Comment by 6cdh** [comment]
Thank you for your patience for a lot of these changes and contribution.

**Comment by amaanq** [comment]
Thanks for merging

File: src/scanner.c (hunk 9 above)
Comment (6cdh) [reply src/scanner.c:16]: The language allows it. Even if the language grammar is very crazy and out of control (it's not here), the parser should do it.

The thinking of details and the pursuit of high-quality code makes a different programmer.

What would Linus do if he's here?
